src/lib/trace/trace.go

```diff
@@ -0,0 +1,139 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package trace
+
+import (
+	"context"
+	"time"
+
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/exporters/jaeger"
+	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracehttp"
+	"go.opentelemetry.io/otel/propagation"
+	"go.opentelemetry.io/otel/sdk/resource"
+	tracesdk "go.opentelemetry.io/otel/sdk/trace"
+	semconv "go.opentelemetry.io/otel/semconv/v1.4.0"
+	oteltrace "go.opentelemetry.io/otel/trace"
+
+	"github.com/goharbor/harbor/src/lib/log"
+	"github.com/goharbor/harbor/src/pkg/version"
+)
+
+func initExporter(ctx context.Context) (tracesdk.SpanExporter, error) {
+	var err error
+	var exp tracesdk.SpanExporter
+	cfg := GetGlobalConfig()
+	if len(cfg.Jaeger.Endpoint) != 0 {
+		// Jaeger collector exporter
+		log.Infof("init trace provider jaeger collector on %s with user %s", cfg.Jaeger.Endpoint, cfg.Jaeger.Username)
+		exp, err = jaeger.New(jaeger.WithCollectorEndpoint(
+			jaeger.WithEndpoint(cfg.Jaeger.Endpoint),
+			jaeger.WithUsername(cfg.Jaeger.Username),
+			jaeger.WithPassword(cfg.Jaeger.Password),
+		))
+	} else if len(cfg.Jaeger.AgentHost) != 0 {
+		// Jaeger agent exporter
+		log.Infof("init trace provider jaeger agent on %s", cfg.Jaeger.AgentHost)
+		exp, err = jaeger.New(jaeger.WithAgentEndpoint(
+			jaeger.WithAgentHost(cfg.Jaeger.AgentHost),
+			jaeger.WithAgentPort(cfg.Jaeger.AgentPort),
+		))
+	} else if len(cfg.Otel.Endpoint) != 0 {
+		// Otel exporter
+		log.Infof("init trace provider otel on %s/%s", cfg.Otel.Endpoint, cfg.Otel.URLPath)
+		opts := []otlptracehttp.Option{
+			otlptracehttp.WithEndpoint(cfg.Otel.Endpoint),
+			otlptracehttp.WithURLPath(cfg.Otel.URLPath),
+			otlptracehttp.WithTimeout(time.Duration(cfg.Otel.Timeout) * time.Second),
+		}
+		if cfg.Otel.Compression {
+			opts = append(opts, otlptracehttp.WithCompression(otlptracehttp.GzipCompression))
+		}
+		if cfg.Otel.Insecure {
+			opts = append(opts, otlptracehttp.WithInsecure())
+		}
+		exp, err = otlptracehttp.New(ctx, opts...)
+	} else {
+		log.Fatalf("Trace is enabled but no tracer provider is specified")
+	}
+	return exp, err
+}
+
+func initProvider(exp tracesdk.SpanExporter) (*tracesdk.TracerProvider, error) {
+	cfg := GetGlobalConfig()
+
+	// prepare attribute resources
+	attriSlice := []attribute.KeyValue{
+		semconv.ServiceNameKey.String(cfg.ServiceName),
+	}
+	if len(version.ReleaseVersion) != 0 {
+		attriSlice = append(attriSlice, semconv.ServiceVersionKey.String(version.ReleaseVersion))
+	}
+	if cfg.Namespace != "" {
+		attriSlice = append(attriSlice, semconv.ServiceNamespaceKey.String(cfg.Namespace))
+	}
+	if cfg.Attributes != nil {
+		for i, a := range cfg.Attributes {
+			attriSlice = append(attriSlice, attribute.String(i, a))
+		}
+	}
+
+	// prepare tp options
+	ops := make([]tracesdk.TracerProviderOption, 0, 4)
```

Comment:
I am not quite sure that the interval and message size can be set when to init a provider? and it's not exposed to the end-user. 

We can discuss it, but this will not impact this PR.

Reply:
I think we can discuss it later. I think we've already introduced a lot of new configs. if a lot of user needs it when they deployed it in production we can add them in the future.